source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,1136 @@
+
+    // big floating point
+    assert(binomial!real(1.2e10, 2.3).approxEqual(5.66794e22, 0.0001));
+    // TODO: rounding errors seem to sum up here
+    //assert(binomial!real(2.245e15, 5.7).approxEqual(7.68342e84, 0.1));
+}
+
+/**
+Lazily computes all _permutations of `r` using $(WEB
+en.wikipedia.org/wiki/Heap%27s_algorithm, Heap's algorithm).
+
+While generating a new item is in `O(k)` (amortized `O(1)`),
+the number of permutations is `|r|!`.
+
+Returns:
+A forward range the elements of which are an $(XREF range,
+indexed) view into `r`.
+*/
+auto permutations(Range, Allocator)(Range r, auto ref Allocator alloc)
+if (isRandomAccessRange!Range && hasLength!Range)
+{
+    return Permutations!(Range)(r, alloc);
+}
+
+/// ditto
+auto permutations(Range)(Range r)
+if (isRandomAccessRange!Range && hasLength!Range)
+{
+    return Permutations!Range(r);
+}
+
+private struct Permutations(Range)
+{
+    private size_t[] indices, state;
+    private Range r;
+    private bool _empty;
+
+    private void _init()
+    {
+        _empty = r.length == 0;
+    }
+
+    this(Range r)
+    {
+        import std.range : iota;
+        import std.array : array;
+
+        this.r = r;
+        size_t rLength = r.length;
+
+        state = new size_t[rLength - 1];
+        indices = iota(size_t(rLength)).array;
+
+        _init();
+    }
+
+    this(Allocator)(Range r, ref Allocator alloc)
+    {
+        import std.range : iota;
+        import std.experimental.allocator: makeArray;
+
+        this.r = r;
+        size_t rLength = r.length;
+
+        state = alloc.makeArray!size_t(rLength - 1);
+        indices = alloc.makeArray!size_t(rLength);
+        auto iotaResult = iota(rLength);
+        foreach(i, ref el; indices)
+        {
+            el = iotaResult[i];
+        }
+        alloc = alloc;
+
+        _init();
+    }
+
+    void dispose(Allocator)(Allocator alloc)
+    {
+        alloc.dispose(state);
+        alloc.dispose(indices);
+    }
+
+    @property auto front()
+    {
+        import std.range : indexed;
+        return r.indexed(indices);
+    }
+
+    void popFront()
+    {
+        import std.algorithm.mutation : swapAt;
+        for (size_t h = 0;;h++)
+        {
+            if (h+2 > indices.length)
+            {
+                _empty = true;
+                break;
+            }
+
+            if (h & 1)
+                indices.swapAt(0, h+1);
+            else
+                indices.swapAt(state[h], h+1);
+
+            if (state[h] == h+1)
+            {
+                state[h] = 0;
+                continue;
+            }
+            state[h]++;
+            break;
+        }
+    }
+
+    @property bool empty()
+    {
+        return _empty;
+    }
+
+    @property typeof(this) save()
+    {
+        typeof(this) c = this;
+        c.indices = indices.dup;
+        c.state = state.dup;
+        return c;
+    }
+
+    @property typeof(this) save(Allocator)(ref Allocator alloc)
```

Comment:
`save` is the standard method in Phobos to save the state of a struct / function. Imho it should work when using an Allocator too - otherwise a lot of the range algorithms will not work. Most of them expect a `ForwardRange`.
https://dlang.org/library/std/range/interfaces/forward_range.html